### ISSUE TITLE
PubSub nightmare: "No Cognito Identity pool provided for unauthenticated access"

### ISSUE THREAD

**Issue by githubflub** [aws-amplify/amplify-js#6244]
There's nothing on google about this issue, so I'm documenting it. I was trying to set up PubSub. Did everything right - all the setup on AWS, using the AWS CLI. Here's my Amplify related code. 

```
import Amplify from '@aws-amplify/core'
import Auth from '@aws-amplify/auth'
import PubSub from '@aws-amplify/pubsub'
window.LOG_LEVEL = 'VERBOSE';
import { AWSIoTProvider } from '@aws-amplify/pubsub/lib/Providers'
Amplify.register(PubSub);

export function configureAmplify() {
   Amplify.configure({
      Auth: {
         region: TS_AWS_REGION,
         userPoolId: USER_POOL_ID,
         userPoolWebClientId: USER_POOL_CLIENT_ID,
         identityPoolId: IDENTITY_POOL_ID,
      }
   })

   Amplify.addPluggable(new AWSIoTProvider({
      aws_pubsub_region: TS_AWS_REGION,
      aws_pubsub_endpoint: IOT_ENDPOINT,
   }))

   PubSub.subscribe('ATopic').subscribe({
      next: data => console.log('Message received', data),
      error: error => console.error(error),
      close: () => console.log('Done'),
    });
}
```
But in the midst of the PubSub.subscribe.subscribe call, there would be a mysterious Unhandled Exception error with the following message "No Cognito Identity pool provided for unauthenticated access". While adding console logs everywhere to try to figure out what was wrong, suddenly, Amplify stopped crashing at the first error and continued until another error showed up "AMQJS0007E Socket error:undefined". I also cleared my node_modules folder so I could switch between using 'aws-amplify' and '@aws-amplify/core'. 

Because I did the setup perfectly, these errors should not have been happening. After many many hours of struggle, I figured out the cause: the folder structure in the node_modules folder. Apparently, I had a few different versions of @aws-amplify packages in my node_modules folder. 

I carefully changed my Amplify-related dependencies to these specific versions. I checked the package.jsons of each of these and according to their dependencies, these are the versions meant to be used together. 

```
    "@aws-amplify/auth": "3.3.0",
    "@aws-amplify/core": "3.4.1",
    "@aws-amplify/pubsub": "3.0.19",
```
After a clean npm install and then deleting the @aws-amplify folder in node_modules, a second npm install produced a folder structure that fixed the errors. I'm showing the structure below. See that there are no node_modules folders in the "auth" or "cache" folders and that in the node_modules folders in "core" and "pubsub", there are no "circular references" to other @aws-amplify packages. 

![image](https://user-images.githubusercontent.com/18061353/86681263-415f3b00-bfb4-11ea-81cb-35407f31c4cd.png)
![image](https://user-images.githubusercontent.com/18061353/86681378-5e940980-bfb4-11ea-9174-8c9e22cdd9f8.png)

I'm not sure if I'm out of the woods yet, because I'm getting two of the following logs when I think I should be expecting just one. I see two zen-observative-ts which is kind of concerning, but at the very least, I've solved a very crazy issue. 

![image](https://user-images.githubusercontent.com/18061353/86682210-31942680-bfb5-11ea-9353-0b71817f12ea.png)

I hope this manages to help you all.

**Comment by githubflub** [comment]
Closed because this is resolved.

**Comment by dingdongping** [comment]
Can PubSub be used alone without relying on @ AWS amplify / auth

**Comment by daniel-bravo** [comment]
I did face the same problem. Solution was (as described by @githubflub): 
1. Remove the pub-sub npm
2. Remove aws-amplify aws-amplify-react-native uuid (I've added those npm following a workshop guide)

Then:
1. Add pub-sub npm. It's important to add it BEFORE aws-amplify aws-amplify-react-native uuid
2. Add aws-amplify aws-amplify-react-native uuid. Also, it's worthy to take a look at this documentation to set the proper permissions at IAM: https://docs.amplify.aws/lib/pubsub/getting-started/q/platform/js

**Comment by billychan** [comment]
We justed encoutered a similar issue and reasoned it about.

The problem may arise if you are using sub modules one by one like '@aws-amplify/auth', '@aws-amplify/pubsub'(in our case it's Analytics). 

The reason is those sub modules could depend on different version of core, when that happens, the authentication happened in one core module and the usage happens on another, makes the second module unauthenticasted.

This line make Auth authenticate the Amplify it depends https://github.com/aws-amplify/amplify-js/blob/aws-amplify%403.3.14/packages/auth/src/Auth.ts#L2149, which is an instance of Amplify not necessary same as what's you imported.

The fix we did are:
1. Carefully check versions of all modules, use pinned specific version for each, all depends on same version of core.
2. Remove node_modules and re-run `yarn install`, check generated yarn.lock to verify there is no duplicated version of core there.

Then it should work.
